## Import Libraries and Dataset 

In [1]:
import numpy as np
import pandas as pd

In [2]:
import os

In [3]:
os.chdir('D:/Pro/02-Artificial-Intelligence-Course/01-DataScienceWithPython/02-Projects/Project-amazonmoviesandtvratings')

In [4]:
pwd

'D:\\Pro\\02-Artificial-Intelligence-Course\\01-DataScienceWithPython\\02-Projects\\Project-amazonmoviesandtvratings'

In [5]:
amazonData = pd.read_csv('Amazon - Movies and TV Ratings.csv')

## Data Understanding

In [6]:
amazonData.shape

(4848, 207)

In [7]:
amazonData.columns

Index(['user_id', 'Movie1', 'Movie2', 'Movie3', 'Movie4', 'Movie5', 'Movie6',
       'Movie7', 'Movie8', 'Movie9',
       ...
       'Movie197', 'Movie198', 'Movie199', 'Movie200', 'Movie201', 'Movie202',
       'Movie203', 'Movie204', 'Movie205', 'Movie206'],
      dtype='object', length=207)

In [8]:
amazonData.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1   AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2  A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3   AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4  A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

In [9]:
amazonData.describe()

Movie1  Movie2  Movie3  Movie4     Movie5  Movie6  Movie7  Movie8  \
count     1.0     1.0     1.0     2.0  29.000000     1.0     1.0     1.0   
mean      5.0     5.0     2.0     5.0   4.103448     4.0     5.0     5.0   
std       NaN     NaN     NaN     0.0   1.496301     NaN     NaN     NaN   
min       5.0     5.0     2.0     5.0   1.000000     4.0     5.0     5.0   
25%       5.0     5.0     2.0     5.0   4.000000     4.0     5.0     5.0   
50%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
75%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
max       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   

       Movie9  Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
count     1.0      1.0  ...  5.000000       2.0       1.0  8.000000  3.000000   
mean      5.0      5.0  ...  3.800000       5.0       5.0  4.625000  4.333333   
std       NaN      NaN  ...  1.643168       0.0       NaN  0.517549  1.154701   
min       5.0      5.0  ...  1.000000       5.0       5.0  4.000000  3.000000   
25%       5.0      5.0  ...  4.000000       5.0       5.0  4.000000  4.000000   
50%       5.0      5.0  ...  4.000000       5.0       5.0  5.000000  5.000000   
75%       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   
max       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   

       Movie202  Movie203  Movie204   Movie205   Movie206  
count  6.000000       1.0  8.000000  35.000000  13.000000  
mean   4.333333       3.0  4.375000   4.628571   4.923077  
std    1.632993       NaN  1.407886   0.910259   0.277350  
min    1.000000       3.0  1.000000   1.000000   4.000000  
25%    5.000000       3.0  4.750000   5.000000   5.000000  
50%    5.000000       3.0  5.000000   5.000000   5.000000  
75%    5.000000       3.0  5.000000   5.000000   5.000000  
max    5.000000       3.0  5.000000   5.000000   5.000000  

[8 rows x 206 columns]

In [10]:
amazonData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4848 entries, 0 to 4847
Columns: 207 entries, user_id to Movie206
dtypes: float64(206), object(1)
memory usage: 7.7+ MB


In [11]:
len(amazonData['user_id'].unique())

4848

## Data Understanding Report :

1) There are 4848 rows and 207 columns

2) There are 4848 unique users

------

In [12]:
ColNullValueCount = amazonData.isna().sum()

In [13]:
ColsToSort = list(amazonData.columns)
ColsToSort.pop(0)

'user_id'

-----

## EXPLORATORY DATA ANALYSIS

## Which movies have maximum views/ratings?

In [14]:
movies = amazonData.drop('user_id',axis=1)

movies.count().sort_values(ascending=False).head(10)

Movie127    2313
Movie140     578
Movie16      320
Movie103     272
Movie29      243
Movie91      128
Movie92      101
Movie89       83
Movie158      66
Movie108      54
dtype: int64

----

## What is the average rating for each movie? 
## Define the top 5 movies with the maximum ratings.

In [15]:
amazonData[ColsToSort].mean().sort_values(ascending=False).head(5)

Movie1      5.0
Movie55     5.0
Movie131    5.0
Movie132    5.0
Movie133    5.0
dtype: float64

In [16]:
avgRatings = amazonData[ColsToSort].mean() 
print("Number of Movies with Average Rating equal to 5.0 : ",avgRatings[avgRatings == 5.0].count())
print("Number of Movies with Average Rating greater than 5.0 : ",avgRatings[avgRatings > 5.0].count())
print("Number of Movies with Average Rating lesser than 5.0 : ",avgRatings[avgRatings < 5.0].count())

Number of Movies with Average Rating equal to 5.0 :  103
Number of Movies with Average Rating greater than 5.0 :  0
Number of Movies with Average Rating lesser than 5.0 :  103


---

## Define the top 5 movies with the least audience

In [17]:
movieCount = movies.count().sort_values(ascending=True)

In [18]:
movieCount.head()

Movie1      1
Movie71     1
Movie145    1
Movie69     1
Movie68     1
dtype: int64

In [19]:
print("The least number of views/audience for a movie is 1")
print("The total number of movies with 1 view/audience is : ", movieCount[movieCount == 1 ].count())

The least number of views/audience for a movie is 1
The total number of movies with 1 view/audience is :  89


---

## Recommendation Model

In [20]:
movies.head()

Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  Movie8  Movie9  \
0     5.0     5.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
1     NaN     NaN     2.0     NaN     NaN     NaN     NaN     NaN     NaN   
2     NaN     NaN     NaN     5.0     NaN     NaN     NaN     NaN     NaN   
3     NaN     NaN     NaN     5.0     NaN     NaN     NaN     NaN     NaN   
4     NaN     NaN     NaN     NaN     5.0     NaN     NaN     NaN     NaN   

   Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  Movie202  \
0      NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
1      NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
2      NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
3      NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
4      NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   

   Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN  

[5 rows x 206 columns]

In [26]:
from surprise import Reader
from surprise import accuracy
from surprise import Dataset
from surprise.model_selection import train_test_split

In [22]:
amazonData.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1   AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2  A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3   AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4  A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

In [24]:
melt_movies = amazonData.melt(id_vars = amazonData.columns[0],        \
                              value_vars = amazonData.columns[1:],     \
                              var_name = "movie_name",                  \
                              value_name = "rating")

In [25]:
melt_movies

user_id movie_name  rating
0       A3R5OBKS7OM2IR     Movie1     5.0
1        AH3QC2PC1VTGP     Movie1     NaN
2       A3LKP6WPMP9UKX     Movie1     NaN
3        AVIY68KEPQ5ZD     Movie1     NaN
4       A1CV1WROP5KTTW     Movie1     NaN
...                ...        ...     ...
998683  A1IMQ9WMFYKWH5   Movie206     5.0
998684  A1KLIKPUF5E88I   Movie206     5.0
998685   A5HG6WFZLO10D   Movie206     5.0
998686  A3UU690TWXCG1X   Movie206     5.0
998687   AI4J762YI6S06   Movie206     5.0

[998688 rows x 3 columns]

In [38]:
reader = Reader(rating_scale=(-1,10))

data = Dataset.load_from_df(melt_movies.fillna(0), reader=reader)

In [39]:
trainSet,testSet = train_test_split(data, test_size=0.25)

In [40]:
trainSet

In [41]:
from surprise import SVD

In [42]:
algo = SVD()

In [43]:
algo.fit(trainSet)

In [44]:
predict = algo.test(testSet)

In [45]:
accuracy.rmse(predict)

RMSE: 0.2793


0.27926397100239103

In [48]:
user_id = 'A3R5OBKS7OM2IR'
movie_id = 'Movie1'
r_ui = 5.0

algo.predict(user_id,movie_id,r_ui, verbose=True)

user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 0.07   {'was_impossible': False}


Prediction(uid='A3R5OBKS7OM2IR', iid='Movie1', r_ui=5.0, est=0.07207917314450574, details={'was_impossible': False})

In [49]:
from surprise.model_selection import cross_validate

In [51]:
cross_validate(algo, data, measures=['RMSE','MAE'],cv=3,verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.2849  0.2794  0.2809  0.2817  0.0023  
MAE (testset)     0.0429  0.0426  0.0424  0.0426  0.0002  
Fit time          30.64   30.76   30.60   30.66   0.07    
Test time         2.84    2.11    2.10    2.35    0.35    


{'test_rmse': array([0.28492352, 0.27943598, 0.28085115]),
 'test_mae': array([0.04285246, 0.04259029, 0.04243195]),
 'fit_time': (30.636741399765015, 30.757975101470947, 30.599949598312378),
 'test_time': (2.840334892272949, 2.1091370582580566, 2.1000728607177734)}

In [52]:
def repeat(algo_type, frame, min_, max_):
    reader = Reader(rating_scale=(min_,max_))
    
    data = Dataset.load_from_df(frame,reader=reader)
    
    algo = algo_type
    
    print(cross_validate(algo,data, measures=['RMSE','MAE'], cv=3, verbose=True))
    
    print("-"*10)
    
    
    user_id = 'A3R5OBKS7OM2IR'
    movie_id = 'Movie1'
    r_ui = 5.0
    print(algo.predict(user_id,movie_id,r_ui, verbose=True))
    
    print("-"*10)
    
    print()

In [53]:
df = movies.iloc[:1212,:50]

melt_df = df.melt(id_vars=df.columns[0],value_vars=df.columns[1:],var_name="movie_name",value_name="rating")

In [54]:
repeat(SVD(),melt_df.fillna(0),-1,10)

repeat(SVD(),melt_df.fillna(melt_df.mean()),-1,10)

repeat(SVD(),melt_df.fillna(melt_df.median()),-1,10)

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.4323  0.4569  0.4543  0.4478  0.0111  
MAE (testset)     0.0873  0.0923  0.0867  0.0887  0.0025  
Fit time          1.65    1.61    1.61    1.62    0.02    
Test time         0.09    0.09    0.09    0.09    0.00    
{'test_rmse': array([0.43228683, 0.45691172, 0.45433547]), 'test_mae': array([0.08726127, 0.0922586 , 0.08671266]), 'fit_time': (1.650956392288208, 1.6062064170837402, 1.6128687858581543), 'test_time': (0.08677554130554199, 0.09122443199157715, 0.08704257011413574)}
----------
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 0.05   {'was_impossible': False}
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 0.05   {'was_impossible': False}
----------

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.0923  0.0885  0.0843  0.0884  0.0033  
MAE (te

In [55]:
from surprise.model_selection import GridSearchCV

In [56]:
param_grid = {'n_epochs':[20,30],
             'lr_all':[0.005,0.01],
             'n_factors':[50,100]}

In [59]:
gs = GridSearchCV(SVD,param_grid, measures=['rmse','mae'], cv=3)

gs.fit(data)

In [60]:
gs.best_score

{'rmse': 0.2772684734279281, 'mae': 0.03984954709365934}